src/pip/_internal/resolution/resolvelib/requirements.py

```diff
@@ -40,10 +41,19 @@ def is_satisfied_by(self, candidate: Candidate) -> bool:
 
 
 class SpecifierRequirement(Requirement):
-    def __init__(self, ireq: InstallRequirement) -> None:
+    def __init__(
+        self,
+        ireq: InstallRequirement,
+        *,
+        drop_extras: bool = False,
```

Comment:
I don’t particularly like using a flag to switch between logic like this; maybe this logic should be in a classmethod (`SpecifierRequirement.without_extras()`?) or subclass (`SpecifierWithoutExtrasRequirement`?)